packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -688,154 +861,585 @@ class _SortData {
+      // common directionality to be able to group into sane groups for the
+      // given directionality, since rectangles can overlap and give different
+      // results for different directionalities.
+      mergeSort<_ReadingOrderSortData>(inBandOfTop, compare: (_ReadingOrderSortData a,_ReadingOrderSortData b) {
+        switch(nearestCommonDirectionality) {
+          case TextDirection.ltr:
+            return a.rect.left.compareTo(b.rect.left);
+          case TextDirection.rtl:
+            return b.rect.right.compareTo(a.rect.right);
+        }
+        assert(false, 'Unhandled directionality $nearestCommonDirectionality');
+        return 0;
+      });
+
+      // Collect the top band into internally sorted groups with shared directionality.
+      final List<_ReadingOrderDirectionalGroupData> bandGroups = collectDirectionalityGroups(inBandOfTop);
+      if (bandGroups.length == 1) {
+        // There's only one directionality group, so just send back the first
+        // one in that group, since it's already sorted.
+        return bandGroups.first.members.first;
       }
-      return topmost;
+
+      // Sort the groups based on the common directionality and bounding boxes.
+      mergeSort<_ReadingOrderDirectionalGroupData>(bandGroups, compare: (_ReadingOrderDirectionalGroupData a, _ReadingOrderDirectionalGroupData b) {
+        switch(nearestCommonDirectionality) {
+          case TextDirection.ltr:
+            return a.rect.left.compareTo(b.rect.left);
+          case TextDirection.rtl:
+            return b.rect.right.compareTo(a.rect.right);
+        }
+        assert(false, 'Unhandled directionality $nearestCommonDirectionality');
+        return 0;
+      });
+      return bandGroups.first.members.first;
     }
 
-    final List<_SortData> data = <_SortData>[
-      for (final FocusNode node in nodes) _SortData(node),
+    final List<_ReadingOrderSortData> data = <_ReadingOrderSortData>[
+      for (final FocusNode node in descendants) _ReadingOrderSortData(node),
     ];
 
     // Pick the initial widget as the one that is leftmost in the band of the
     // topmost, or the topmost, if there are no others in its band.
-    final List<_SortData> sortedList = <_SortData>[];
-    final List<_SortData> unplaced = data.toList();
-    _SortData current = pickFirst(unplaced);
+    final List<_ReadingOrderSortData> sortedList = <_ReadingOrderSortData>[];
+    final List<_ReadingOrderSortData> unplaced = data.toList();
+    _ReadingOrderSortData current = pickNext(unplaced);
     sortedList.add(current);
     unplaced.remove(current);
 
     while (unplaced.isNotEmpty) {
-      final _SortData next = pickFirst(unplaced);
+      final _ReadingOrderSortData next = pickNext(unplaced);
       current = next;
       sortedList.add(current);
       unplaced.remove(current);
     }
-    return sortedList.map((_SortData item) => item.node);
+    return sortedList.map<FocusNode>((_ReadingOrderSortData item) => item.node);
```

Comment:
Make the sortedList of type `List<FocusNode>` so you can avoid this last map?

Reply:
Good call, thanks!